File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static boolean[] visited;
    static ArrayList<Integer>[] arr;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken()); //노드 수
        int M = Integer.parseInt(st.nextToken()); //엣지 수
        int V = Integer.parseInt(st.nextToken()); //시작점
        arr = new ArrayList[N+1];
        //객체 생성
        for (int i = 1; i <= N; i++) {
            arr[i] = new ArrayList<>();
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int node1 = Integer.parseInt(st.nextToken());
            int node2 = Integer.parseInt(st.nextToken());

            arr[node1].add(node2);
            arr[node2].add(node1);
        }
        //번호가 작은 노드부터 방문하기 위해 정렬
        for (int i = 1; i <= N; i++) {
            Collections.sort(arr[i]);
        }
        visited = new boolean[N + 1];
        DFS(V);
        System.out.println();
        visited = new boolean[N + 1];
        BFS(V);
        System.out.println();
        br.close();
    }
    private static void DFS(int Node) {
        //방문 순서를 출력해야 하므로 처음에 출력
        System.out.print(Node + " ");
        //방문 노드는 표시
        visited[Node] = true;
        //재귀로 stack이 빌때까지 반복
        for(int i : arr[Node]) {
            if(!visited[i]) DFS(i);
        }
    }
    private static void BFS(int Node) {
        //BFS는 큐로 구현
        Queue<Integer> queue = new LinkedList<>();
        queue.add(Node);
        //방문 노드는 표시
        visited[Node] = true;
        //큐가 빌때까지 반복
        while (!queue.isEmpty()) {
            int nowNode = queue.poll();
            System.out.print(nowNode + " ");
            
            for(int i : arr[nowNode]) {
                if(!visited[i]) {
                    //큐에 방문한 노드는 체크
                    visited[i] = true;
                    queue.add(i);
                }
            }
        }
    }
}

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
from collections import deque

def dfs(V):
    global visited, graph

    if visited[V]:
        return
    visited[V] = True

    print(V, end=' ')

    for i in graph[V]:
        dfs(i)

def bfs(V):
    global visited, graph

    q = deque()
    q.append(V)
    visited[V] = True

    while q:
        cur = q.popleft()
        print(cur, end=' ')

        for i in graph[cur]:
            if not visited[i]:
                q.append(i)
                visited[i] = True

N, M, V = map(int, input().split())

graph = [[] for _ in range(N + 1)]

for _ in range(M):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

for i in range(N + 1):
    graph[i].sort()

visited = [False] * (N + 1)
dfs(V)
print()

visited = [False] * (N + 1)
bfs(V)
print()
